### PR TITLE
Add tpch as a compiler test

### PR DESCRIPTION
**Pull request by mihaibudiu**
TPCH as a benchmark has surfaced a couple of bugs in the compiler; this PR fixes the bugs and enables TPCH to be compiled into legal Rust.

### CODE REVIEW

File: sql-to-dbsp-compiler/SQL-compiler/out14840349038327279478.tmp
Comment (gz) [inline sql-to-dbsp-compiler/SQL-compiler/out14840349038327279478.tmp:1]: Probably accidental addition?

**Review by gz — APPROVED** [review]

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/TpchTest.java [hunk 1]
```diff
@@ -0,0 +1,853 @@
...
+import org.dbsp.sqlCompiler.circuit.DBSPCircuit;
+import org.dbsp.sqlCompiler.compiler.backend.rust.RustFileWriter;
+import org.dbsp.util.Utilities;
+import org.junit.Test;
+
+import java.io.IOException;
+
+public class TpchTest extends BaseSQLTests {
+    static final String tpch = "CREATE TABLE NATION  ( N_NATIONKEY  INTEGER NOT NULL,\n" +
+            "                            N_NAME       CHAR(25) NOT NULL,\n" +
```
Comment (lalithsuresh) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/TpchTest.java:12]: Could you make this a test resources file?

**Review by lalithsuresh — APPROVED** [review]

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/TpchTest.java [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+            "    avg(l_discount) as avg_disc,\n" +
+            "    count(*) as count_order\n" +
+            "from\n" +
+            "    lineitem\n" +
+            "where\n" +
+            "    l_shipdate <= date '1998-12-01' - interval '71' DAY\n" +
+            "group by\n" +
+            "    l_returnflag,\n" +
+            "    l_linestatus\n" +
+            "order by\n" +
```
Comment (mohsinbeg) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/TpchTest.java:114]: Order By is not incremental, so you can remove it. Or have a copy of the view without it that we'll run.

**Review by mohsinbeg — APPROVED** [review]
Is there a plan to run the TpchTest queries with data?

**Comment by mihaibudiu** [comment]
If we can have data of reasonable size and reference outputs, sure.
We haven't implemented topK yet, which is groupby + limit. That's an important one, which is more incremental in nature.

**Comment by ryzhyk** [comment]
> If we can have data of reasonable size and reference outputs, sure. We haven't implemented topK yet, which is groupby + limit. That's an important one, which is more incremental in nature.

Does calcite compile group-by + limit into top-k or do you need to infer that?

**Comment by mihaibudiu** [comment]
Don't recall having seen such an operator, but perhaps it's not too difficult. We have to design it for DBSP first, and it has to support flexible sorting.

**Comment by ryzhyk** [comment]
It exists in DBSP. Flexible sorting probably requires wrapper types with overloaded comparison operator. And that's another thing that will require different implementations in Rust and JIT.

**Pull request merged**